File: macos/SeedSeeker/Tests/SeedSeekerKitTests/BlanketRequirementsTests.swift
import Foundation
import XCTest
@testable import SeedSeekerKit

final class BlanketRequirementsTests: XCTestCase {
    private func wand(_ key: Int64, blanket: Bool = false) throws -> ItemRequirement {
        try ItemRequirement(key: key, item: ItemCatalog.findById("wand_frost"), upgrade: 0,
                            kind: .wand, upgradeMatch: .any, blanket: blanket)
    }

    func testBlanketsSurviveDocumentsLinksResultsAndSavedQueries() throws {
        var requirements = try ["wand_lightning", "wand_disintegration", "wand_frost"].enumerated().map { index, id in
            try ItemRequirement(key: Int64(index + 1), item: ItemCatalog.findById(id), upgrade: 2,
                                kind: .wand, upgradeMatch: .atLeast)
        }
        requirements.append(try ItemRequirement(key: 4, item: nil, upgrade: 3, kind: .wand,
                                               source: .wandmakerReward, blanket: true))
        let query = SavedQuery(requirements: requirements)
        let persisted = try JSONDecoder().decode(SavedQuery.self, from: JSONEncoder().encode(query))
        let restored = try ResultsExport.decodeQuery(ResultsExport.encodeQuery(query))
        let linked = try DeepLink.decode(DeepLink.encodeLink(for: query))
        let exported = try ResultsExport.decode(ResultsExport.encode(query, seeds: ["AAA-AAA-AAA"], appVersion: "test")).query
        for decoded in [persisted, restored, linked, exported] {
            XCTAssertEqual(decoded.requirements.map(\.blanket), [false, false, false, true])
            XCTAssertEqual(decoded.requirements.last?.source, .wandmakerReward)
            XCTAssertEqual(decoded.requirements.last?.upgrade, 3)
            XCTAssertNotNil(decoded.validated())
            XCTAssertNoThrow(try SearchRequest(requirements: decoded.requirements))
        }
    }

    func testBlanketsCannotStackOrJoinOrdinaryRequirements() throws {
        let requirements = try [wand(1), wand(2, blanket: true), wand(3, blanket: true)]
        XCTAssertEqual(requirements.boardCount, 3)
        XCTAssertFalse(requirements.canStack(requirements.boardItems()[1]))
        XCTAssertEqual(requirements.setStackCount(requirements.boardItems()[1], 3), requirements)
        XCTAssertEqual(requirements.joinAlternatives(source: 0, target: 1), requirements)
        let grouped = requirements.joinAlternatives(source: 1, target: 2)
        XCTAssertEqual(grouped.boardCount, 2)
        XCTAssertTrue(grouped.allSatisfy { $0.identityGroup == nil })
        XCTAssertNoThrow(try SearchRequest(requirements: grouped))
        let linked = try DeepLink.decode(DeepLink.encodeLink(for: SavedQuery(requirements: grouped)))
        XCTAssertEqual(linked.requirements.filter(\.blanket).count, 2)
        XCTAssertEqual(linked.requirements.slotCount, 2)
        XCTAssertEqual(grouped.detach(1).boardCount, 3)
    }

    func testInvalidBlanketsAreRejectedBeforeSearching() throws {
        XCTAssertThrowsError(try SearchRequest(requirements: [wand(1, blanket: true)]))
        var ordinary = try wand(1); ordinary.alternativeGroup = 1
        var blanket = try wand(2, blanket: true); blanket.alternativeGroup = 1
        XCTAssertThrowsError(try SearchRequest(requirements: [ordinary, blanket]))
        XCTAssertThrowsError(try ItemRequirement(key: 1, item: nil, upgrade: 0, kind: .wand,
                                               upgradeMatch: .any, identityGroup: 1, blanket: true))
    }
}

File: windows/SeedSeeker.Tests/BlanketRequirementsTests.cs
using System.Text.Json;
using Xunit;

namespace SeedSeeker.Tests;

public sealed class BlanketRequirementsTests
{
    private static ItemRequirement Wand(bool blanket = false) => new()
    {
        Kind = ItemKind.Wand, Item = ItemCatalog.Find("wand_frost"), Blanket = blanket,
    };

    [Fact]
    public void BlanketsSurviveDocumentsLinksResultsAndSettings()
    {
        var query = new QuerySettings { Requirements = new(new[] { "wand_lightning", "wand_disintegration", "wand_frost" }
            .Select(id => new ItemRequirement { Kind = ItemKind.Wand, Item = ItemCatalog.Find(id), Upgrade = 2, UpgradeMatch = UpgradeMatch.AtLeast })) };
        query.Requirements.Add(new() { Kind = ItemKind.Wand, Blanket = true, Upgrade = 3,
            UpgradeMatch = UpgradeMatch.Exactly, Source = ScoutItemSource.WandmakerReward });
        var document = ResultsExport.EncodeQueryDocument(query);
        Assert.Contains("\"blanket\":true", document);
        var link = NativeEngine.TryEncodeShareLink(document);
        Assert.NotNull(link);
        foreach (var restored in new[] {
            ResultsExport.DecodeQueryDocument(document),
            ResultsExport.DecodeQueryDocument(NativeEngine.TryDecodeShareText(link!)!),
            ResultsExport.Decode(ResultsExport.Encode(query, ["AAA-AAA-AAA"], "test")).Query,
            JsonSerializer.Deserialize<QuerySettings>(JsonSerializer.Serialize(query))!, query.Clone(),
        })
        {
            Assert.Equal(new[] { false, false, false, true }, restored.Requirements.Select(r => r.Blanket));
            Assert.Equal(3, restored.Requirements.Last().Upgrade);
            Assert.Equal(ScoutItemSource.WandmakerReward, restored.Requirements.Last().Source);
            Assert.Null(QueryRelationships.Validate(restored));
        }
    }

    [Fact]
    public void BlanketsStaySeparateFromStacksAndOrdinaryAlternatives()
    {
        var requirements = new[] { Wand(), Wand(true), Wand(true) };
        var board = QueryRelationships.BoardItems(requirements);
        Assert.Equal(3, board.Count);
        Assert.False(QueryRelationships.CanStack(requirements, board[1]));
        Assert.Equal(3, QueryRelationships.SetStackCount(requirements, board[1], 3).Count);
        Assert.Equal(requirements, QueryRelationships.JoinAlternatives(requirements, 0, 1));
        var grouped = QueryRelationships.JoinAlternatives(requirements, 1, 2);
        Assert.Equal(2, QueryRelationships.BoardCount(grouped));
        Assert.All(grouped, r => Assert.Null(r.IdentityGroup));
        var query = new QuerySettings { Requirements = new(grouped) };
        Assert.Null(QueryRelationships.Validate(query));
        Assert.NotNull(NativeEngine.TryEncodeShareLink(ResultsExport.EncodeQueryDocument(query)));
        Assert.Equal(3, QueryRelationships.BoardCount(QueryRelationships.Detach(grouped, 1)));
    }

    [Fact]
    public void InvalidBlanketsAreRejectedBeforeSearching()
    {
        Assert.NotNull(QueryRelationships.Validate(new() { Requirements = [Wand(true)] }));
        var ordinary = Wand(); ordinary.AlternativeGroup = 1;
        var blanket = Wand(true); blanket.AlternativeGroup = 1;
        Assert.NotNull(QueryRelationships.Validate(new() { Requirements = [ordinary, blanket] }));
        blanket.AlternativeGroup = null; blanket.IdentityGroup = 1;
        Assert.NotNull(QueryRelationships.Validate(new() { Requirements = [Wand(), blanket] }));
        blanket.IdentityGroup = null; blanket.SelectTrinket = true;
        Assert.NotNull(QueryRelationships.Validate(new() { Requirements = [Wand(), blanket] }));
    }

    [Fact]
    public void ConflictingBlanketStopsBeforeScanningAndReportsImpossible()
    {
        var query = new QuerySettings { Requirements = new(new[] { "wand_lightning", "wand_disintegration" }
            .Select(id => new ItemRequirement { Kind = ItemKind.Wand, Item = ItemCatalog.Find(id), Upgrade = 2, UpgradeMatch = UpgradeMatch.Exactly })) };
        query.Requirements.Add(new() { Kind = ItemKind.Wand, Blanket = true, Upgrade = 3,
            UpgradeMatch = UpgradeMatch.Exactly });
        using var search = new NativeEngine().StartResumed(query, 42, 1_000, 1);
        Assert.True(SpinWait.SpinUntil(() => search.Status().State != SearchState.Running,
            TimeSpan.FromSeconds(5)));
        var status = search.Status();
        Assert.True(status.IsImpossibleQuery);
        Assert.Equal(0, status.Scanned);
        Assert.Empty(search.Poll(1));
        Assert.Equal((42L, 1_000L), search.ResumeHint());
        Assert.False(new SearchStatus(SearchState.Completed, 0, 0, 0, 0).IsImpossibleQuery);
        Assert.False(new SearchStatus(SearchState.Cancelled, 0, 1_000, 0, 0).IsImpossibleQuery);
        Assert.False(new SearchStatus(SearchState.Completed, 1_000, 1_000, 0, 0).IsImpossibleQuery);
    }
}
